src/install.rs

```diff
@@ -278,6 +287,373 @@ wiping them with `wipefs -a`.\n"
+                "fedora-coreos" => get_fcos_live_urls(
+                    &booted_deployment
+                        .base_metadata
+                        .fedora_coreos_stream
+                        .unwrap(),
+                    &booted_deployment.version,
+                    &booted_deployment.base_metadata.basearch,
+                    config.fetch_retries,
+                )?,
+                "rhcos" => get_rhcos_live_urls(
+                    &booted_deployment.version,
+                    &booted_deployment.base_metadata.basearch,
+                    config.fetch_retries,
+                )?,
+                x => bail!("invalid CoreOS variant {}", x),
+            }
+        };
+
+        eprintln!("Downloading {}...", &initramfs_url);
+        let mut initrd =
+            download_live_artifact(&initramfs_url, config.insecure, config.fetch_retries)?;
+        if config.defer_rootfs {
+            deferred_rootfs_url = Some(rootfs_url);
+        } else {
+            eprintln!("Downloading {}...", &rootfs_url);
+            let mut rootfs =
+                download_live_artifact(&rootfs_url, config.insecure, config.fetch_retries)?;
+            concat_initrds(&mut initrd, &mut rootfs)?;
+        }
+        initrd
+    };
+
+    // build kargs
+    let mut kargs = String::new();
+    add_karg(&mut kargs, "coreos.inst.install_dev", &config.dest_device);
+    //kargs.push_str("console=tty0 console==ttyS0");
+    add_karg(&mut kargs, "console", &config.console);
+    if let Some(stream) = config.stream {
+        add_karg(&mut kargs, "coreos.inst.stream", &stream);
+    }
+    if let Some(url) = config.image_url {
+        add_karg(&mut kargs, "coreos.inst.image_url", url.as_str());
+    }
+    if let Some(url) = config.ignition_url {
+        add_karg(&mut kargs, "coreos.inst.ignition_url", url.as_str());
+    }
+    if let Some(id) = config.platform {
+        add_karg(&mut kargs, "coreos.inst.platform_id", &id);
+    }
+    if !config.save_partlabel.is_empty() {
+        add_karg(
+            &mut kargs,
+            "coreos.inst.save_partlabel",
+            &config.save_partlabel.join(","),
+        );
+    }
+    if !config.save_partindex.is_empty() {
+        add_karg(
+            &mut kargs,
+            "coreos.inst.save_partindex",
+            &config.save_partindex.join(","),
+        );
+    }
+    if config.insecure {
+        kargs.push_str(" coreos.inst.insecure");
+    }
+    if config.skip_reboot {
+        kargs.push_str(" coreos.inst.skip_reboot");
+    }
+    if let Some(url) = deferred_rootfs_url {
+        add_karg(&mut kargs, "coreos.live.rootfs_url", url.as_str());
+    };
+
+    eprintln!("Loading kernel and initramfs with arguments: {}", &kargs);
+    runcmd!(
+        "kexec",
+        "--load",
+        &kernel,
+        "--initrd",
+        fd_path(&final_initrd),
+        "--append",
+        &kargs
+    )?;
+
+    eprintln!("Pivoting");
+    runcmd!("systemctl", "kexec")?;
```

Comment:
kexec is great but I think can sometimes have bugs, so it might be useful to think about a mode that instead writes the kernel/initrd to e.g. `/boot/coreos-installer-reprovision/{kernel,initramfs.img}` and updates the bootloader entries to use them manually.  Then we can do the reinstall from the initramfs on the next boot.

The hit is a double reboot, but it avoids kexec.

Reply:
Yeah, possible. One thing worth mentioning though is that this strategy could be useful not just for reinstalling CoreOS from CoreOS, but any situation in which you're running on top of the same block device you'd like to install to. So I could imagine this being used in the container workflow too. In which case, not having to fiddle with boot configuration is a plus because you don't know what the setup is (and of course `reinstall` doesn't really make sense in that context; maybe would belong better under `install` with a switch?).